### ISSUE TITLE
Feature Request: Option to resume training with the same number of steps remaining, at the same point in the learning rate schedule

### ISSUE THREAD

**Issue by morphinapg** [d8ahazard/sd_dreambooth_extension#1101]
Currently, if you hit cancel on training, and then start again, you have exactly the same number of steps remaining as if you had just started, and the learning rate schedule starts over as well. Not to say that training completely starts over, because it does start from a model that's already been partially trained, but the schedule does start over, which has a different effect than resuming at the exact same point. While I'm sure the current way it works can be beneficial to some people, I would imagine it would make more sense to work the same way it does in the webui's training tab for embeddings and hypernetworks, where it resumes exactly where you left off. So if you're halfway through training and you press cancel, you would resume and continue the remaining half of your training in the next session. 

While this could be somewhat simulated by changing the number of steps, getting rid of warmup, and changing the learning rate to where it left off, it wouldn't be perfect. Obviously, this is extra work that really shouldn't be necessary, but also, if you have a curved learning rate schedule like cosine for example, then it will not resume at the same position along the curve either. 

Perhaps many people in here train for a small enough number of steps that this wouldn't be necessary, but if you're trying to train many concepts, and/or a large data set, there's a good chance it would need to be training for a long time and if you're like me, you may need your GPU for other purposes during some of that time.

So I think it would be good to add an option to be able to pause and resume at the exact point you left off. That way those who like the way it works now could still make use of that, but those like me would have an option to continue our training at the same point we left off.

**Comment by saunderez** [comment]
I am working on OOM resume that could also work if deliberately stopped. Don't think resuming with the same LR is a good idea though, that's really rough on the model. My plan was to check if a warmup enabled scheduler is in use, and if so add that many steps to the total steps remaining to allow gradual return to the high LR.

**Comment by morphinapg** [comment]
> I am working on OOM resume that could also work if deliberately stopped. Don't think resuming with the same LR is a good idea though, that's really rough on the model. My plan was to check if a warmup enabled scheduler is in use, and if so add that many steps to the total steps remaining to allow gradual return to the high LR.

Why would it be rough on the model? Wouldn't it be no different than if you had just continued the next step?

I understand why it's rough when first starting training, but that shouldn't be an issue if you're already partly through your training.

The point of this would be to resume exactly as if you had continued the next step instead of stopping. So learning rate is at the same point along the curve (including warmup steps if you stopped during warmup) and has exactly the same number of steps remaining. 

Basically like, if you stopped on step 14250, then it would resume at step 14250, rather than starting over at step 1. Since learning rate calculation is determined based on which step # you're on, simply resuming at the correct step number _should_ achieve the effect I was asking for. The total number of steps wouldn't increase either.

Basically, just use lifetime steps as the current step #

**Comment by saunderez** [comment]
Because currently resuming is restarting. It's a new instance of the
optimiser, as far as it's concerned it's the first time it's seen the model
so needs to adjust to it again.

For a true resume you'd need to save the optimiser state. I believe that
would allow you to resume as if nothing had happened, this is how the A1111
TI training does it.

Investigating further is on my to-do list, I need to sort out my OOM resume
and saving optimiser state could possibly kill two birds with the one stone.

**Comment by morphinapg** [comment]
> Because currently resuming is restarting. It's a new instance of the optimiser, as far as it's concerned it's the first time it's seen the model so needs to adjust to it again. For a true resume you'd need to save the optimiser state. I believe that would allow you to resume as if nothing had happened, this is how the A1111 TI training does it. Investigating further is on my to-do list, I need to sort out my OOM resume and saving optimiser state could possibly kill two birds with the one stone.

Hmm. I guess I don't really understand what the optimizer does. I've actually simulated resuming by adjusting the learning rate to where it was, removing warmup steps, and changing total epochs to match how many epochs I have left to do. From my experiments, this seems to work just fine with certain learning rate schedules like polynomial or linear for example, but obviously wouldn't work the same way for cosine. Comparing it to sessions where I was able to train the full steps in one session, the results didn't seem to be any worse. 

But I mean sure if you can come up with something that does the same thing the A1111 TI training works then that would be great.

**Comment by morphinapg** [comment]
While it's not a total solution, I did come up with a workaround on Windows, better than what I suggested for now. If you open Resource Monitor and find the python.exe that is currently running, you can right click it and say suspend process. After this you are free to use the GPU with whatever else you want, games, rendering, even other instances of A1111. When you resume the process, you may get an error message in the console about its connection to the UI, but it will continue training just fine, so it looks like all data in memory, including optimizer state, gets restored. One time when I had to use my GPU heavily while training was paused, when I went to resume the process, my computer locked up for a few minutes and I could hear my hard drive being accessed. So it sounds like when your GPU needs to use the VRAM with another process, the data that was being used by that python.exe is saved to a page file, and restored when the process is resumed, so suspending the process doesn't even prevent you from using the VRAM for other things, which is especially helpful. 

Like I said, not a total fix, but for people using windows, it may be enough until something better gets implemented.
